# Dataset Inventory — Financial Manager AI

## Objective

This notebook performs an initial inventory of the datasets considered for the Financial Manager AI project.

The purpose of this stage is to understand the structure, quality, granularity, and potential applicability of each dataset before performing data cleaning, transformation, modeling, or database integration.

The following aspects will be evaluated:

- dataset structure;
- number of records and attributes;
- data types;
- missing values;
- attribute cardinality;
- possible identifiers;
- dimensions and measures;
- temporal coverage;
- record granularity;
- potential relationships between attributes;
- applicability to the Financial Manager AI modules.

No permanent data cleaning or transformation will be performed at this stage.

## 1. Imports

Import the libraries required for the initial dataset inspection.

In [1]:
import pandas as pd

## 2. Dataset Loading

The dataset analyzed in this section is the **Receiver General Accounting Transactions** dataset.

The original raw file is loaded without modifying its values, columns, or structure.    

In [2]:
df = pd.read_csv(
    "../data/raw/Receiver General Accounting Transactions/tcrg-rgat.csv"
)

### Dataset Metadata

- **Dataset:** Receiver General Accounting Transactions
- **File:** `tcrg-rgat.csv`
- **Format:** CSV
- **Source:** Kaggle
- **Original data:** preserved without modification
- **Documentation / Data Dictionary:** [link or reference, if available]

## 3. General Dataset Structure

This section provides an initial overview of the dataset, including its number of rows, number of columns, column names, data types, memory usage, and non-null values.

This information helps identify the overall structure of the dataset before performing more detailed analysis.

In [3]:
df.shape

(9282, 16)

In [4]:
df.columns.tolist()

['Accounting-Control-Number-Numéro-contrôle-comptable',
 'Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal',
 'Fiscal-Year-Année-Fiscale',
 'Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal',
 "Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable",
 'DepartmentNumber-Numéro-de-Ministère',
 'Fiscal-Month-Mois-Fiscal',
 "Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-la-pièce-de-journal",
 'Financial-Coding-Block-Bloc-de-codage-financier',
 'Subleger-Account-Identifier-Compte-du-grand-livre-auxiliaire',
 'Credit/Debit-Code-Code-Crédit/Débit',
 "Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal",
 'Control-Data-Type-Type-de-données-de-contrôle',
 'Accounting-Summary-Transaction-Line-Count-Sommaire-comptable-transaction-compte-ligne',
 'GBS-Control-Number-Numéro-de-contrôle-du-SBG',
 'General-Ledger-Account-Code-Code-du-compte-du-grand-livre-général']

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9282 entries, 0 to 9281
Data columns (total 16 columns):
 #   Column                                                                                 Non-Null Count  Dtype  
---  ------                                                                                 --------------  -----  
 0   Accounting-Control-Number-Numéro-contrôle-comptable                                    9282 non-null   int64  
 1   Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal                             9256 non-null   str    
 2   Fiscal-Year-Année-Fiscale                                                              9282 non-null   str    
 3   Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal                       178 non-null    float64
 4   Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable                           9282 non-null   str    
 5   DepartmentNumber-Numéro-de-Ministère                                                   

### Initial Observations

The dataset contains 9,282 records and 16 attributes.

The columns include integer, floating-point, and string data types.

Some attributes contain substantial amounts of missing data, while several core accounting fields appear to be complete.

## 4. Data Sample and Record Granularity

A small sample of records is inspected to better understand the meaning of the attributes and determine what a single row represents.

The dataset is transposed for visualization purposes because the original column names are long and difficult to inspect horizontally.

In [6]:
print(df.head().T.to_string())

                                                                                                       0                  1                  2                  3                  4
Accounting-Control-Number-Numéro-contrôle-comptable                                         202309300033       202309300034       202309300035       202309300036       202309300037
Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal                                         JVEFI              JVEFI              JVEFI              JVEFI              JVEFI
Fiscal-Year-Année-Fiscale                                                                      2023/2024          2023/2024          2023/2024          2023/2024          2023/2024
Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal                                     NaN                NaN                NaN                NaN                NaN
Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable                                  2

### Preliminary Granularity

Each row appears to represent an individual accounting journal voucher item.

This interpretation is based on fields such as:

- Journal Voucher Item Identifier;
- Credit/Debit Code;
- Journal Voucher Item Amount;
- Accounting Effective Date;
- accounting and subledger identifiers.

This assumption should later be validated against the official dataset documentation or data dictionary.

## 5. Data Types

This section examines the data type assigned to each attribute by Pandas.

Data types provide useful technical information, but they do not necessarily indicate the business meaning of a field.

For example, a department identifier may be stored as a numeric value while still representing a categorical dimension rather than a quantitative measure.

In [7]:
df.dtypes

Accounting-Control-Number-Numéro-contrôle-comptable                                        int64
Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal                                   str
Fiscal-Year-Année-Fiscale                                                                    str
Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal                         float64
Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable                                 str
DepartmentNumber-Numéro-de-Ministère                                                     float64
Fiscal-Month-Mois-Fiscal                                                                   int64
Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-la-pièce-de-journal            int64
Financial-Coding-Block-Bloc-de-codage-financier                                              str
Subleger-Account-Identifier-Compte-du-grand-livre-auxiliaire                             float64
Credit/Debit-Code-Code-Crédit/

## 6. Missing Values

Missing-value analysis is used to identify attributes with incomplete coverage.

Both the absolute number and percentage of missing values are evaluated.

At this stage, columns are not removed. The purpose is only to document their completeness and determine whether they may be useful in later stages of the project.

In [8]:
missing = pd.DataFrame(
    {
        "missing_count": df.isna().sum(),
        "missing_percent": (df.isna().mean() * 100).round(2),
    }
)

missing.sort_values("missing_percent", ascending=False)

,missing_count,missing_percent
Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal,9104,98.08
General-Ledger-Account-Code-Code-du-compte-du-grand-livre-général,9091,97.94
Financial-Coding-Block-Bloc-de-codage-financier,9000,96.96
Subleger-Account-Identifier-Compte-du-grand-livre-auxiliaire,2186,23.55
GBS-Control-Number-Numéro-de-contrôle-du-SBG,239,2.57
Control-Data-Type-Type-de-données-de-contrôle,178,1.92
DepartmentNumber-Numéro-de-Ministère,26,0.28
Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal,26,0.28
Accounting-Summary-Transaction-Line-Count-Sommaire-comptable-transaction-compte-ligne,26,0.28
Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-la-pièce-de-journal,0,0.00


### Preliminary Observations

Some attributes have extremely low coverage, including fields with more than 95% missing values.

These fields may have limited analytical usefulness, but they will not be removed during the inventory stage.

Other attributes, including accounting dates, fiscal periods, debit/credit classification, and transaction amounts, appear to have significantly better coverage.

## 7. Attribute Cardinality

Cardinality represents the number of distinct values contained in each attribute.

This analysis helps identify:

- possible identifiers;
- categorical dimensions;
- low-cardinality classification fields;
- continuous or highly variable measures.

High cardinality does not automatically mean that a column is a unique identifier. Possible keys must be validated separately.

In [9]:
df.nunique().sort_values(ascending=False)

Accounting-Control-Number-Numéro-contrôle-comptable                                      9134
Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal                     8276
GBS-Control-Number-Numéro-de-contrôle-du-SBG                                             8214
Subleger-Account-Identifier-Compte-du-grand-livre-auxiliaire                              354
Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable                              188
Accounting-Summary-Transaction-Line-Count-Sommaire-comptable-transaction-compte-ligne      78
DepartmentNumber-Numéro-de-Ministère                                                       54
Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal                                 39
Control-Data-Type-Type-de-données-de-contrôle                                              34
Journal-Voucher-Identifier-Identificateur-de-la-pièce-de-journal                           30
Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-

### Preliminary Observations

Some attributes have a number of distinct values close to the total number of records, suggesting that they may function as identifiers or control numbers.

Other attributes contain only a few dozen distinct values, suggesting categorical dimensions such as department, journal voucher type, or control type.

Potential identifiers will be tested for uniqueness later in the notebook.

## 8. Distribution of Key Categorical Attributes

This section examines the frequency distribution of selected categorical attributes.

The objective is to understand the composition of the dataset and identify dominant categories, rare categories, and potentially useful dimensions for financial analysis.

In [10]:
df["Credit/Debit-Code-Code-Crédit/Débit"].value_counts(dropna=False)

Credit/Debit-Code-Code-Crédit/Débit
DR    7231
CR    2051
Name: count, dtype: int64

In [11]:
df["Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal"].value_counts(
    dropna=False
)

Journal-Voucher-Type-Code-Code-du-type-de-pièce-de-journal
ADEPT    3439
JVEFI    2289
JVFIR     563
DOMFR     386
JVIVI     188
EPAYR     187
JVBDP     187
JVRDI     187
JVRGI     187
RDBOA     187
JVLVP     187
RGEPA     186
JVIVR     178
JVFID     176
GCMO      148
JVBRT     134
JVFRD     120
JVEFR      74
AFDEP      61
JVRFI      48
NaN        26
RDPCM      22
CGAGR      16
RGEIC      11
CADJU      11
JVBRC      10
UNSPD      10
FMSPM       9
RMSCM       9
MDDEP       9
RDSDT       7
BNKFE       6
LVTSR       5
CRCPC       4
GCRCD       4
FORFR       3
SUSPD       2
RCDSP       2
RCDCP       2
CITYE       2
Name: count, dtype: int64

In [12]:
df["Fiscal-Year-Année-Fiscale"].value_counts(dropna=False)

Fiscal-Year-Année-Fiscale
2023/2024    9225
2022/2023      57
Name: count, dtype: int64

In [13]:
df["DepartmentNumber-Numéro-de-Ministère"].value_counts(dropna=False)

DepartmentNumber-Numéro-de-Ministère
122.0    2390
97.0     2101
14.0     1436
18.0     1318
6.0       860
5.0       456
30.0       82
163.0      51
35.0       50
127.0      50
165.0      49
41.0       39
86.0       34
7.0        31
119.0      27
1.0        26
NaN        26
148.0      25
130.0      19
79.0       17
87.0       16
85.0       15
34.0       15
22.0       14
33.0       14
95.0       14
21.0       13
50.0       13
54.0        9
47.0        7
67.0        7
136.0       7
56.0        6
191.0       5
37.0        4
88.0        4
11.0        4
124.0       3
135.0       3
53.0        3
144.0       2
61.0        2
9.0         2
52.0        2
62.0        1
139.0       1
142.0       1
27.0        1
170.0       1
13.0        1
195.0       1
25.0        1
96.0        1
180.0       1
123.0       1
Name: count, dtype: int64

In [14]:
df["Control-Data-Type-Type-de-données-de-contrôle"].value_counts(dropna=False)

Control-Data-Type-Type-de-données-de-contrôle
COK    3825
CMC    2289
JMT     563
JMZ     208
CMD     188
JMR     187
JMY     187
CMI     187
JMG     187
JND     187
JML     187
NaN     178
CMF     178
JMU     176
CMT     134
JMV     120
CME      74
COL      64
JMW      48
COH      16
JNC      11
COF      11
CMJ      10
CMA      10
JOD       9
JOE       9
COJ       9
JNE       7
COP       6
JMS       5
COE       4
CMB       2
CNU       2
CNS       2
JOB       2
Name: count, dtype: int64

### Preliminary Observations

The dataset contains both debit and credit accounting entries.

Journal voucher types show several distinct accounting categories, although some types are considerably more frequent than others.

Most records are concentrated in a single fiscal year, which may limit the dataset's usefulness for long-term time-series analysis and forecasting.

## 9. Temporal Coverage

Temporal coverage is evaluated to determine the actual period represented by the dataset.

This is especially important for financial trend analysis, anomaly detection, seasonality analysis, and forecasting.

The accounting effective date is converted to a datetime representation for analysis without modifying the original raw dataset.

In [15]:
dates = pd.to_datetime(
    df["Accounting-Effective-Date-Date-d'entrée-en-vigueur-comptable"], errors="coerce"
)

In [16]:
dates.isna().sum()

np.int64(0)

In [17]:
dates.min(), dates.max()

(Timestamp('2023-04-03 00:00:00'), Timestamp('2024-01-03 00:00:00'))

In [18]:
print(df["Fiscal-Month-Mois-Fiscal"].value_counts().sort_index().to_string())

Fiscal-Month-Mois-Fiscal
1      961
2     1058
3     1021
4      793
5     1109
6     1003
7     1132
8     1079
9     1009
10      60
12      57


### Temporal Coverage Assessment

The available time range should be evaluated to determine whether the dataset provides sufficient historical depth for each Financial Manager AI capability.

A dataset may be suitable for descriptive analysis or anomaly detection while still being insufficient for reliable long-term forecasting.

### Preliminary Observations

The accounting effective dates range from April 3, 2023 to January 3, 2024.

The dataset therefore contains less than one full year of observed accounting activity.

Most observations are distributed across the available fiscal months, although the number of records is not uniform across periods.

Because the dataset uses fiscal-month identifiers rather than calendar-month names, the relationship between fiscal months and calendar months should be confirmed using the official dataset documentation.

The relatively short historical coverage may be sufficient for descriptive financial analysis and some anomaly-detection experiments, but it may limit the reliability of long-term time-series forecasting.

## 10. Financial Measure Analysis

The main monetary attribute is analyzed to understand its statistical distribution.

The objective is to identify the scale, variability, central tendency, and extreme values of accounting transaction amounts.

Extreme values are not automatically considered errors because large financial transactions may be legitimate.

In [19]:
df["Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal"].describe()

count    9.282000e+03
mean     3.298092e+08
std      1.887057e+09
min      0.000000e+00
25%      5.782925e+02
50%      1.767720e+03
75%      6.725283e+04
max      3.955579e+10
Name: Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal, dtype: float64

### Preliminary Observations

The transaction amount distribution is highly right-skewed.

The median transaction amount is approximately 1,767.72, while the mean is substantially higher due to a relatively small number of extremely large accounting entries.

The maximum observed transaction amount is approximately 39.56 billion.

This suggests the presence of strong outliers or legitimately large accounting transactions. These observations should not be removed automatically and will require further investigation during the EDA and anomaly-detection stages.

Because of the highly skewed distribution, the median may provide a more representative measure of a typical transaction than the arithmetic mean.





### Debit and Credit Comparison

Transaction amounts are also analyzed separately by debit and credit classification.

This provides a clearer understanding of the monetary behavior of the two accounting entry types.

In [20]:
df.groupby("Credit/Debit-Code-Code-Crédit/Débit")[
    "Journal-Voucher-Item-Amount-Montant-de-l'item-de-la-pièce-de-journal"
].agg(["count", "sum", "mean", "median", "min", "max"])

,count,sum,mean,median,min,max
Credit/Debit-Code-Code-Crédit/Débit,,,,,,
CR,2051,5.188250e+10,2.529620e+07,1723.06,0.0,6.184370e+09
DR,7231,3.009406e+12,4.161812e+08,1785.92,0.0,3.955579e+10


### Debit and Credit Observations

Debit entries are considerably more frequent than credit entries in the dataset.

Although the median transaction amounts for debit and credit entries are relatively similar, debit entries contain substantially larger extreme values.

The total debit amount is also significantly higher than the total credit amount in this dataset.

These differences should not be interpreted directly as expenses versus revenues, because debit and credit classifications represent accounting entry directions rather than business income and expense categories.

The result suggests that debit and credit transactions may need to be analyzed separately during anomaly detection and financial analysis.

## 11. Duplicates and Potential Keys

This section evaluates complete duplicate rows and investigates columns that may function as identifiers.

A column with high cardinality is not automatically unique.

In accounting datasets, a control number may identify a journal entry containing multiple individual items, which means that a combination of attributes may be required to uniquely identify each record.

In [21]:
df.duplicated().sum()

np.int64(0)

In [22]:
df["Accounting-Control-Number-Numéro-contrôle-comptable"].duplicated().sum()

np.int64(148)

### Duplicate Analysis

No complete duplicate rows were identified in the dataset.

However, the Accounting Control Number contains 148 duplicated occurrences, confirming that this attribute cannot uniquely identify individual records by itself.

This result motivates the investigation of a composite key.

### Key Investigation

If the accounting control number is repeated across multiple records, the dataset may require a composite key.

Possible combinations of control numbers, journal voucher identifiers, and journal voucher item identifiers should be investigated before defining the final database schema.

### Composite Key Investigation

The Accounting Control Number is not unique by itself.

This test evaluates whether combining the Accounting Control Number with the Journal Voucher Item Identifier produces a unique identifier for each record.

In [23]:
df.duplicated(
    subset=[
        "Accounting-Control-Number-Numéro-contrôle-comptable",
        "Journal-Voucher-Item-Identifier-Identificateur-de-l'item-de-la-pièce-de-journal",
    ]
).sum()

np.int64(0)

### Composite Key Result

The combination of Accounting Control Number and Journal Voucher Item Identifier produced zero duplicate records.

This indicates that the two attributes together uniquely identify each row in the current dataset.

Therefore, this combination is a strong candidate for a composite key.

The final database key should still be confirmed during the schema-design stage and validated against the official dataset documentation.

## 12. Attribute Classification

The attributes are classified according to their likely analytical role.

The main categories used are:

- **Identifier:** used to distinguish records, transactions, accounts, or control entities;
- **Dimension:** used to categorize, filter, group, or segment financial information;
- **Measure:** quantitative value that can be aggregated or statistically analyzed;
- **Control Field:** operational or accounting metadata used for validation, traceability, or internal control.

This classification is preliminary and may change after reviewing the official data dictionary.

| Attribute | Preliminary Classification | Description |
|---|---|---|
| Accounting Control Number | Identifier / Control Field | Accounting record control reference |
| Journal Voucher Type Code | Dimension | Type of accounting journal entry |
| Fiscal Year | Temporal Dimension | Fiscal year associated with the entry |
| Journal Voucher Identifier | Identifier | Journal voucher reference |
| Accounting Effective Date | Temporal Dimension | Effective accounting date |
| Department Number | Organizational Dimension | Department or organizational unit |
| Fiscal Month | Temporal Dimension | Fiscal accounting month |
| Journal Voucher Item Identifier | Identifier | Individual journal voucher item |
| Financial Coding Block | Accounting Dimension | Financial classification or coding information |
| Subledger Account Identifier | Accounting Dimension | Subledger account reference |
| Credit/Debit Code | Financial Dimension | Indicates debit or credit |
| Journal Voucher Item Amount | Measure | Monetary value of the journal item |
| Control Data Type | Control Dimension | Classification used for accounting control |
| Accounting Summary Transaction Line Count | Control Measure | Number of transaction lines |
| GBS Control Number | Identifier / Control Field | System control reference |
| General Ledger Account Code | Accounting Dimension | General ledger account classification |

## 13. Applicability to Financial Manager AI

The dataset is evaluated against the six core capabilities planned for the Financial Manager AI V1.

The objective is not to force a single dataset to support every module.

Instead, each dataset should be used only for the capabilities that are supported by its actual attributes, granularity, temporal coverage, and business meaning.

| Financial Manager Capability | Preliminary Assessment | Reason |
|---|---|---|
| Accounts Payable | Inadequate | Supplier, invoice, due date, payment status, and outstanding balance information are not available |
| Financial Data Analysis | Suitable | Contains dates, monetary values, departments, accounting categories, and debit/credit information |
| Budget Variance | Inadequate / Partial | No explicit budget or planned amount is available |
| Financial Document RAG | Inadequate | This dataset contains structured accounting records rather than financial documents |
| Financial Anomaly Detection | Potentially Suitable | Contains a relatively large number of accounting observations, monetary values, categories, accounts, and organizational dimensions |
| Cash Flow Forecast | Partial | Accounting journal entries do not necessarily represent actual cash inflows and outflows |

### Preliminary Dataset Role

The Receiver General Accounting Transactions dataset appears to be most useful for:

- financial data analysis;
- accounting transaction exploration;
- debit and credit analysis;
- analysis by department or account;
- potential financial anomaly detection.

Additional datasets will likely be required for Accounts Payable, Budget Variance, and Cash Flow Forecast capabilities.

## 14. Preliminary Conclusions

The Receiver General Accounting Transactions dataset is primarily an accounting journal transaction dataset.

Its structure provides useful information for financial analysis through accounting dates, fiscal periods, departments, account identifiers, debit/credit classifications, and monetary transaction amounts.

The dataset appears particularly relevant to the Financial Data Analysis module and may also provide useful observations for Financial Anomaly Detection.

However, it does not contain sufficient information to independently support Accounts Payable, Budget Variance, or complete Cash Flow Forecast capabilities.

Several attributes have extremely low data coverage and will require further evaluation during the data cleaning and schema-design stages.

The dataset also has limited historical coverage, with most records concentrated in a single fiscal year. This limitation should be considered before using the dataset for time-series forecasting.

No attributes will be permanently removed or transformed during this inventory stage.

The next step is to perform the same inventory process on the **Company Financials Dataset** and compare both datasets before defining the Financial Manager AI operational data model.